src/Client/State/DCC.hs

```
@@ -0,0 +1,348 @@
+            if size == 0
+              then return Nothing
+              else return $ Just (fromIntegral size)
+
+insertAsNewMax :: DCCOffer -> DCCState -> DCCState
+insertAsNewMax newoffer (DCCState offers transfers) =
+  let newmax = if I.null offers then 1 else succ . fst $ I.findMax offers
+      newOffers = I.insert newmax newoffer offers
+  in DCCState newOffers transfers
+
+-- TODO: maybe replace with some lens magic?
+ctcpToTuple :: IrcMsg -> Maybe (UserInfo, Identifier, Text, Text)
+ctcpToTuple (Ctcp fromU target command txt) =
+  Just (fromU, target, command, txt)
+ctcpToTuple (CtcpNotice fromU target command txt) =
+  Just (fromU, target, command, txt)
+ctcpToTuple _ = Nothing
+
+-- | Check the status of a download at @Key@ by checking the invariants
+--   at @DCCState@
+statusAtKey :: Key -> DCCState -> ConnectionStatus
+statusAtKey key (DCCState offers transfers)
+  | Nothing <- lookupOffer = NotExist
+  | (Just _) <- lookupOffer, Nothing <- I.lookup key transfers = Pending
+  | otherwise = _dtStatus (transfers I.! key)
+  where
+    lookupOffer = I.lookup key offers
+
+-- | Craft a CTCP message indicating we wan to resume a download at the
+--   offset
+resumeMsg :: Word32 {- offset -}
+          -> DCCOffer
+          -> (String, String) -- (target, txt)
+resumeMsg sizeoffset offer =
+  let filename = _dccFileName offer
+      port = show (_dccPort offer)
+      sizeoffset' = show sizeoffset
+      quoting = if any (== ' ') filename then "\"" else mempty
+
+      txt = "RESUME" <> " " <> quoting <> filename <> quoting <> " " <> port
+            <> " " <> sizeoffset'
+      target = Text.unpack $ views uiNick idText (_dccFromInfo offer)
+  in (target, txt)
+
+-- | Modify the @DCCState@ following the corresponding @DCCUpdate@
+acceptUpdate :: DCCUpdate -> DCCState -> DCCState
+acceptUpdate (Accept k port offset) state =
+  case view (dsOffers . at k) state of
+    Nothing -> error "Not a valid key"
```

Could this kill the client if something went wrong? Could we silently ignore a missing key instead? What would the consequences be?

---

We could simply ignore it. This function came to be as some code that was too big to be included directly on `doDCCUpdate`, so the checking that that `case` was not `Nothing` it's done there (and it's the only place where `acceptUpdate` is called. I put an `error` so playing on ghci would show that something was obviously wrong, but I will change that to "do nothing".